executor/executor.go

```diff
@@ -616,25 +616,193 @@ func init() {
 type ProjectionExec struct {
 	baseExecutor
 
-	evaluatorSuit    *expression.EvaluatorSuit
+	evaluatorSuit    *expression.EvaluatorSuit // dispatched to projection workers.
 	calculateNoDelay bool
+
+	prepared              bool
+	projectionConcurrency int
+	output                chan *projectionOutput
+	fetcher               projectionInputFetcher
+	waitGroup             sync.WaitGroup
+	workers               []*projectionWorker
+}
+
+type projectionInputFetcher struct {
+	child  Executor
+	input  chan *projectionInput
+	output chan *projectionOutput
+}
+
+func (w *projectionInputFetcher) run(ctx context.Context, dst chan<- *projectionOutput) {
+	defer func() {
+		for input := range w.input {
+			close(input.worker.input)
+			close(input.worker.output)
+		}
+	}()
+
+	for {
+		input, ok1 := <-w.input
+		if !ok1 {
+			return
+		}
+
+		output, ok2 := <-w.output
+		if !ok2 {
+			close(input.worker.input)
+			close(input.worker.output)
+			return
+		}
+
+		dst <- output
+
+		err := w.child.Next(ctx, input.chk)
+		if err != nil {
+			output.done <- errors.Trace(err)
+			close(input.worker.input)
+			close(input.worker.output)
+			return
+		}
+
+		if input.chk.NumRows() == 0 {
+			output.done <- nil
+			close(input.worker.input)
+			close(input.worker.output)
+			return
+		}
+
+		input.worker.input <- input
+
+		input.worker.output <- output
+	}
+}
+
+type projectionWorker struct {
+	ctx           sessionctx.Context
+	evaluatorSuit *expression.EvaluatorSuit
+
+	input  chan *projectionInput
+	output chan *projectionOutput
+}
+
+func (w *projectionWorker) run(giveBack chan<- *projectionInput, waitGroup *sync.WaitGroup) {
+	defer waitGroup.Done()
+
+	for {
+		input, ok1 := <-w.input
+		if !ok1 {
+			return
+		}
+
+		output, ok2 := <-w.output
+		if !ok2 {
+			return
+		}
+
+		err := w.evaluatorSuit.Run(w.ctx, input.chk, output.chk)
+		output.done <- errors.Trace(err)
+
+		if err != nil {
+			return
+		}
+		giveBack <- input
+	}
+}
+
+type projectionInput struct {
+	chk    *chunk.Chunk
+	worker *projectionWorker
+}
+
+type projectionOutput struct {
+	chk  *chunk.Chunk
+	done chan error
 }
 
 // Open implements the Executor Open interface.
 func (e *ProjectionExec) Open(ctx context.Context) error {
 	if err := e.baseExecutor.Open(ctx); err != nil {
 		return errors.Trace(err)
 	}
+
+	if e.evaluatorSuit.Vectorizable() {
+		e.projectionConcurrency = 8
```

Comment:
Make it configurable?

Reply:
If this is point query, we do not need to use multiple gotoutines.